doc/services/device_mgmt/smp_groups/smp_group_0.rst

```diff
@@ -669,3 +671,92 @@ where:
     | "rc"         | :ref:`mcumgr_smp_protocol_status_codes`; will not appear   |
     |              | if 0                                                       |
     +--------------+------------------------------------------------------------+
+
+.. _mcumgr_smp_group_0_named_group_discovery:
+
+Named Group Discovery
+*********************
+
+Used to obtain group ID for Named Groups, which are groups that does not have
+to have constant Group ID, as it may depend on group order registration.
```

Comment:
This seems like a huge change, why do we want group orders to not be static? If img_mgmt is defined as 1, it should always be group 1. This will add code bloat on both a zephyr device and MCUmgr client device that does not really have any benefit except longer/slower process and larger flash/RAM usage.

Reply:
The static, already existing groups will not be touched. Also groups that will have system wide usage may have been given static ID in the future.
This is only for user groups to avoid collisions when you try to use SMP groups provided from different sources that use the same ID as base. If everyone does USER_BASE + 1, everyone gets the same group ID.

Reply:
But that's fine if they get the same group ID because vendor A's product will be used with vendor A's software, it does not have to work with vendor B's software and vendor B can use the same ID in their system because their software is designed for their system.

Reply:
For any custom handlers I have done prior, the group ID was a Kconfig with a default value (the default value was never changed) and that same ID was used for every product.

Reply:
Examples: https://github.com/LairdCP/zephyr_lib/blob/main/components/Kconfig.lcz_event_manager#L86
https://github.com/LairdCP/zephyr_lib/blob/main/components/Kconfig.lcz_qrtc#L64
https://github.com/LairdCP/zephyr_lib/blob/main/components/Kconfig.lcz_shell_log#L13

Reply:
You can propose Kconfig template, if that is possible, that can be used to help with avoiding collisions and hardcoding PERUSER groups, but this will still need to be reflected in client applications.

Reply:
The bit I don't understand is what problem this is attempting to resolve. A vendor takes zephyr and adds their handlers, they also create their mobile app, they don't need to worry about collisions. So is this for derivatives of zephyr which are they used by vendors? Even in that case, if they add derivative handlers, the end vendor will start with the zephyr derivatives in what they get already so would start ordering from that point onwards, and again create their apps with their handler IDs decided. I can't see where this system would be used.

Reply:
> The bit I don't understand is what problem this is attempting to resolve. A vendor takes zephyr and adds their handlers, they also create their mobile app, they don't need to worry about collisions. So is this for derivatives of zephyr which are they used by vendors? Even in that case, if they add derivative handlers, the end vendor will start with the zephyr derivatives in what they get already so would start ordering from that point onwards, and again create their apps with their handler IDs decided. I can't see where this system would be used.

This gives you freedom from registering groups whether this is done by you code or 3rdparty library you are linking.
You may be linking binary blob of jar into your APP that serves some of commands, delivered for you by somebody, and the blob will be able to figure out where is the group it supports at run-time.
If you use some group that is only added on demand (debug, some modem update, stats of specific module, instrumenting, whatever) you do not have to assign it static ID and hold several tool configuration to handle all your apps where the group landed in different places, you can just ask the app where the group is.

Reply:
So here is what this would require:

- Clients would need to connect and then have to scan essentially a table for supported commands, storing their IDs in a look up table, this means that the zephyr code now needs to store these extra commands and the string names of every group name
- Application (zephyr) code itself now has to have some way of lookup up what the group IDs are internally so a duplicate group name to ID is also needed
- Notifications/callbacks now have the same requirement, IDs are now completely random so internally on the zephyr side, applications need to query what the IDs are and then store them
- Any external code using notifications/callbacks will now see random events and random data assuming that all events are being logged out via e.g. a UART, and an external application would somehow need to query the notification IDs from the zephyr device which means more query functions and more lookup tables
- This makes the zephyr code and client application needlessly complex, and it is certainly going to boost the flash requirements, I would doubt this code would build for nRF52832 for example any more
- At which point, the question becomes why not just get rid of group IDs altogether and replace them with group names, that would at least give a unified interface for this idea instead of introducing lookup tables and needing a translation array every time a device connects. Changing a user application to use names instead of IDs would be far simply too, the cbor array would be adjusted rather than needing chunks of new code and to rework a state machine

Reply:
> So here is what this would require:
> 
>     * Clients would need to connect and then have to scan essentially a table for supported commands, storing their IDs in a look up table, this means that the zephyr code now needs to store these extra commands and the string names of every group name

Only for named groups. Note that PERUSER range and static groups are left unnamed, which is even stated in changes that they will not be given names. I doubt there will be many groups registered that way, but if they will be it is up to user how much flash for strings can be spared.

> 
>     * Application (zephyr) code itself now has to have some way of looking up what the group IDs are internally so a duplicate group name to ID function is also needed

Same query functions may be used for discovery and internal usage, I see no problem here.

> 
>     * Notifications/callbacks would also have the same requirement with this, IDs would be now completely random so internally on the zephyr side, applications need to query what the IDs are and then store them

Yes, in a named range IDs may be random. You can query them with the same functions as above.

> 
>     * Any external code using notifications/callbacks will now see random events and random data assuming that all events are being logged out via e.g. a UART, and an external application would somehow need to query the notification IDs from the zephyr device which means more query functions and more lookup tables

You can log extra info into log file on group ID assignment and don't have to query for anything.

> 
>     * With the previous point, it also means that log files from different firmware revisions are potentially going to be incompatible and useless without a corresponding translation file for each firmware version

You can log extra info into log file on group ID assignment.

> 
>     * This makes the zephyr code and client application needlessly complex, and it is certainly going to boost the flash requirements, I would doubt this code would build for nRF52832 for example any more

Yes, that is common knowledge that you can not everything on every platform.

> 
>     * At which point, the question becomes why not just get rid of group IDs altogether and replace them with group names, that would at least give a unified interface for this idea instead of introducing lookup tables and needing a translation array every time a device connects. Changing a user application to use names instead of IDs would be far simpler too, the cbor array would be adjusted rather than needing chunks of new code and to rework a state machine

Because this is not intended to replace static groups. This is only provided to allow easily extend apps wit additional groups. The doc states that there will be no names given to static groups.
Static groups are there to stay and will probably be extended with new groups in time, if users start to propose some groups useful for Zephyr.
Static groups are also needed by MCUboot, and I see no way to remove them. MCUboot will never support named groups, because there is no reason.

Reply:
> Yes, in a named range IDs may be random. You can query them with the same functions as above.

Callback groups do not match notification groups as per `MGMT event callback group IDs. Note that this is not a 1:1 mapping with mcumgr_group_t values.` so would need additional functions